### PR TITLE
Adds response header support

### PR DESCRIPTION
**Pull request by jkotalik**

### CODE REVIEW

File: samples/SampleServer/HeaderEncoding.cs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+namespace SampleServer
+{
+    internal static class HeaderEncoding
+    {
+        // It should just be ASCII or ANSI, but they break badly with un-expected values. We use UTF-8 because it's the same for
+        // ASCII, and because some old client would send UTF8 Host headers and expect UTF8 Location responses
+        // (e.g. IE and HttpWebRequest on intranets).
+        private static Encoding Encoding = new UTF8Encoding(encoderShouldEmitUTF8Identifier: false, throwOnInvalidBytes: false);
+
+        internal static unsafe string GetString(sbyte* pBytes, int byteCount)
```
Comment (davidfowl) [inline samples/SampleServer/HeaderEncoding.cs]: Change this to byte* instead of sbyte*

File: samples/SampleServer/IISAwaitable.cs [hunk 2]
```diff
@@ -26,7 +26,8 @@ public class IISAwaitable : ICriticalNotifyCompletion
 
             return NativeMethods.REQUEST_NOTIFICATION_STATUS.RQ_NOTIFICATION_CONTINUE;
         };
-        
+
```
Comment (davidfowl) [inline samples/SampleServer/IISAwaitable.cs]: nit: extra space

File: samples/SampleServer/IISHttpContext.cs [hunk 3]
```diff
@@ -33,6 +34,7 @@ public abstract partial class IISHttpContext
         private readonly GCHandle _readOperationGcHandle;
         private readonly IISAwaitable _readOperation = new IISAwaitable();
         private BufferHandle _inputHandle;
+        private List<GCHandle> _pinnedHeaders;
```
Comment (davidfowl) [inline samples/SampleServer/IISHttpContext.cs]: This doesn't need to be a field.

File: samples/SampleServer/IISHttpContext.cs [hunk 4]
```diff
@@ -135,8 +136,9 @@ public IISHttpContext(PipeFactory pipeFactory, IntPtr pHttpContext)
         public async Task FlushAsync(CancellationToken cancellationToken)
         {
             await InitializeResponse();
```
Comment (davidfowl) [inline samples/SampleServer/IISHttpContext.cs]: Lets just put it all into flush async and remove InitializeResponse

File: samples/SampleServer/IISHttpContext.cs [hunk 5]
```diff
@@ -158,15 +160,126 @@ private Task InitializeResponse()
...
+            }
+            catch (Exception)
+            {
+                FreePinnedHeaders(pinnedHeaders);
+                throw;
+            }
+            return pinnedHeaders;
+        }
+
+        private void FreePinnedHeaders(List<GCHandle> pinnedHeaders)
```
Comment (davidfowl) [inline samples/SampleServer/IISHttpContext.cs]: static

File: samples/SampleServer/IISHttpContext.cs (hunk 3 above)
Comment (jkotalik) [reply samples/SampleServer/IISHttpContext.cs]: How else would you recommend passing that state to FlushAsync?

File: samples/SampleServer/IISHttpContext.cs (hunk 4 above)
Comment (jkotalik) [reply samples/SampleServer/IISHttpContext.cs]: Nvrm I see how now XD

File: samples/SampleServer/HeaderEncoding.cs [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+        // It should just be ASCII or ANSI, but they break badly with un-expected values. We use UTF-8 because it's the same for
+        // ASCII, and because some old client would send UTF8 Host headers and expect UTF8 Location responses
+        // (e.g. IE and HttpWebRequest on intranets).
+        private static Encoding Encoding = new UTF8Encoding(encoderShouldEmitUTF8Identifier: false, throwOnInvalidBytes: false);
+
+        internal static unsafe string GetString(byte* pBytes, int byteCount)
+        {
+            // net451: return new string(pBytes, 0, byteCount, Encoding);
+
+            var charCount = Encoding.GetCharCount((byte*)pBytes, byteCount);
```
Comment (davidfowl) [inline samples/SampleServer/HeaderEncoding.cs]: You can remove the cast now.

File: samples/SampleServer/IISHttpContext.cs [hunk 7]
```diff
@@ -132,15 +137,9 @@ public IISHttpContext(PipeFactory pipeFactory, IntPtr pHttpContext)
...
 
-        public async Task FlushAsync(CancellationToken cancellationToken)
-        {
-            await InitializeResponse();
-
-            // TODO: Call FlushAsync
-        }
-
-        private Task InitializeResponse()
+        public IISAwaitable FlushAsync(CancellationToken cancellationToken, ReadableBuffer buffer)
```
Comment (davidfowl) [inline samples/SampleServer/IISHttpContext.cs]: This doesn't look right. Flush shouldn't take a buffer.

**Pull request merged**